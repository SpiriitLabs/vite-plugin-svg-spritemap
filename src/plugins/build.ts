import type { ExternalOption } from 'rollup'
import type { Plugin, ResolvedConfig } from 'vite'
import type { Options, Pattern } from '../types'
import { posix as path } from 'node:path'
import { getFileName } from '../helpers/filename'
import { SVGManager } from '../svgManager'

export default function BuildPlugin(iconsPattern: Pattern, options: Options): Plugin {
  let config: ResolvedConfig
  let fileRef: string
  let fileName: string
  let svgManager: SVGManager
  const spritemapFilter = new RegExp(`/${options?.route || '__spritemap'}`, 'g')
  const pluginExternal: ExternalOption = new RegExp(`/${options?.route || '__spritemap'}`)

  return <Plugin>{
    name: 'vite-plugin-svg-spritemap:build',
    apply: 'build',
    config(config) {
      const configExternal = config.build?.rollupOptions?.external
      let finalExternal: ExternalOption = pluginExternal

      if (Array.isArray(configExternal)) {
        configExternal.push(pluginExternal)
        finalExternal = configExternal
      }
      else if (typeof configExternal === 'string' || typeof configExternal === 'object') {
        finalExternal = [configExternal, pluginExternal]
      }
      else if (typeof configExternal === 'function') {
        finalExternal = (source, importer, isResolved) => {
          if (source.match(pluginExternal))
            return true

          const res = configExternal(source, importer, isResolved)
          return res
        }
      }

      return {
        build: {
          rollupOptions: {
            external: finalExternal,
          },
        },
      }
    },
    configResolved(_config) {
      config = _config
      svgManager = new SVGManager(iconsPattern, options, config)
    },
    async buildStart() {
      await svgManager.updateAll()

      if (typeof options.output === 'object') {
        fileName = getFileName(
          options.output.filename,
          'spritemap',
          svgManager.spritemap,
          'svg',
        )
        const filePath = path.join(config.build.assetsDir, fileName)
        fileRef = this.emitFile({
          type: 'asset',
          needsCodeReference: false,
          name: options.output.name,
          source: svgManager.spritemap,
          fileName: filePath,
          originalFileName: options.output.name,
        })
      }
    },
    transform(code) {
      if (typeof options.output !== 'object' || !spritemapFilter.test(code))
        return

      // prevent sveltekit rewrite
      const base = config.base.startsWith('.')
        ? config.base.substring(1)
        : config.base

      return {
        code: code.replace(
          spritemapFilter,
          path.join(base, this.getFileName(fileRef)),
        ),
        map: null,
      }
    },
  }
}
