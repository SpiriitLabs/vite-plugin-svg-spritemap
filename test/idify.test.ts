import type { UserOptions } from '../src/types'
import { describe, expect, it } from 'vitest'
import { buildVite } from './helper/build'

const idifyConfigs: Record<string, UserOptions['idify']> = {
  custom: name => `custom-${name}`,
}

describe('idify', () => {
  for (const key in idifyConfigs) {
    if (Object.prototype.hasOwnProperty.call(idifyConfigs, key)) {
      it.concurrent(key, async () => {
        const idify = idifyConfigs[key]
        const result = await buildVite({ name: `idify_${key}`, options: { idify } })
        if (!('output' in result))
          return
        const asset = result.output.find(
          asset =>
            asset.name?.startsWith('spritemap.') && asset.name.endsWith('.svg'),
        )

        expect(asset).toBeDefined()

        if (asset && 'source' in asset)
          expect(asset.source).toMatchSnapshot()
      })
    }
  }
})
