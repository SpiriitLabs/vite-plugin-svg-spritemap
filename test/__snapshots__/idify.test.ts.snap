// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`idify > custom 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><symbol id="sprite-custom-no_viewbox" viewBox="0 0 32 32"><path fill="red" d="M7 13a9 9 0 0 1 18 0v3.807l1.928 4.822A1 1 0 0 1 26 23H6a1 1 0 0 1-.928-1.371L7 16.807zm5.125 12A3.99 3.99 0 0 0 16 28a3.99 3.99 0 0 0 3.875-3z"/></symbol><use xlink:href="#sprite-custom-no_viewbox" width="32" height="32" y="0"/><view id="sprite-custom-no_viewbox-view" viewBox="0 0 32 32"/><symbol id="sprite-custom-no_width_height" viewBox="0 0 32 32"><path fill="red" d="M7 13a9 9 0 0 1 18 0v3.807l1.928 4.822A1 1 0 0 1 26 23H6a1 1 0 0 1-.928-1.371L7 16.807zm5.125 12A3.99 3.99 0 0 0 16 28a3.99 3.99 0 0 0 3.875-3z"/></symbol><use xlink:href="#sprite-custom-no_width_height" width="32" height="32" y="32"/><view id="sprite-custom-no_width_height-view" viewBox="0 32 32 32"/><symbol id="sprite-custom-spiriit" viewBox="0 0 125 41"><path d="M0 25.5h6.2c.4 1.9 2.2 3 4.6 3 2.3 0 3.9-.7 3.9-2.5 0-1.2-.7-2-2.5-2.3l-5.8-1c-4-.7-6-2.6-6-6 0-4.7 4.5-7 9.9-7 5.7 0 9.6 2.6 9.9 7h-6c-.3-1.7-1.7-2.6-4-2.6-2 0-3.6.6-3.6 2.1 0 1 .6 1.7 2.3 2l5.8.9c4.1.7 6.2 2.7 6.2 6.2 0 5.2-4.7 7.6-10.2 7.6S.6 30.4 0 25.5m48-4.2c0 6.4-3.7 11.7-9.5 11.7-3.4 0-6.6-1.6-7.7-7.3v14.9h-6.6V10h5.6l.7 6.6c1.4-5.6 4.7-7 8.3-7 5.5 0 9.2 5.2 9.2 11.7m-6.6 0c0-3.4-2.1-6.4-5.3-6.4-3.3 0-5.3 2.8-5.3 6.4s2 6.3 5.3 6.3c3.2 0 5.3-2.9 5.3-6.3M60 32.6h-6.6V15.2h-3.3V10H60zM53.1 3.9c0-2.2 1.6-3.4 3.6-3.4s3.6 1.1 3.6 3.4c0 2.2-1.6 3.3-3.6 3.3-2.1 0-3.6-1.1-3.6-3.3m27 6.1v5.2h-2.6c-3.9 0-6 2.1-6 6.4v11h-6.6V10h5.4l.6 7c.9-5.2 3.4-7 6.5-7zM93 32.6h-6.6V15.2h-3.3V10H93zM86.1 3.9c0-2.2 1.6-3.4 3.6-3.4s3.6 1.1 3.6 3.4c0 2.2-1.6 3.3-3.6 3.3s-3.6-1.1-3.6-3.3M125 27.1v5.4h-4.7c-4.8 0-7.2-2.5-7.2-7.4v-9.9h-3.3V10h3.3V3.9h6.6v1.3c0 2.3-1.4 4.8-4.9 4.8H125v5.2h-5.2v9c0 1.9 1 2.9 2.8 2.9z"/><path fill="#3960ff" d="M106.6 32.6H100V15.2h-3.3V10h9.9zM99.7 3.9c0-2.2 1.6-3.4 3.6-3.4s3.6 1.1 3.6 3.4c0 2.2-1.6 3.3-3.6 3.3s-3.6-1.1-3.6-3.3"/></symbol><use xlink:href="#sprite-custom-spiriit" width="118" height="38" y="64"/><view id="sprite-custom-spiriit-view" viewBox="0 64 118 38"/><symbol class="iconify iconify--file-icons" id="sprite-custom-vite" viewBox="0 0 510 512"><path d="M493.89 58.275 355.178 83.558 379.282 0 186.79 37.718l-2.999 50.64L15.145 58.214C3.53 57.538-3.238 65.879 1.558 77.46l244.056 427.983c5.253 8.575 17.347 8.91 22.65 0L507.575 77.419c5.4-9.676-2.874-21.018-13.685-19.144m-237 435.435L17.87 74.556l164.993 29.491-7.778 131.365 67.632-15.608-18.858 92.344 51.374-15.608-25.495 123.397c-1.27 8.069 9.241 12.362 14.44.812l150.22-299.792-74.135 14.308 10.086-34.962 140.91-25.684z"/></symbol><use xlink:href="#sprite-custom-vite" width="31.88" height="32" y="102"/><view class="iconify iconify--file-icons" id="sprite-custom-vite-view" viewBox="0 102 31.88 32"/></svg>"`;
